### PR TITLE
BUG: `line terminator` and '\n to \r\n' problem in Windows(Issue #20353)

### LINKED ISSUE #20353: DataFrame.to_csv not using correct line terminator value
*Method 1:*
```
        data_frame.to_csv(file_path, sep=self.delimiter, float_format='%.2f',
                                  index=False, line_terminator='\n')

```
*Method 2:*
```
        with open(file_path, mode='w', newline='\n') as f:
            data_frame.to_csv(f, sep=self.delimiter, float_format='%.2f',
                              index=False)
```
#### Problem description

I noticed a strange behavior when using *pandas.DataFrame.to_csv* method on Windows (pandas version 0.20.3). When calling the method using _method 1_ with a file path, it's creating a new file using the \r line terminator,  I had to use method two to make it work.

### PR DESCRIPTION
**Pull request by deflatSOCO**
- [ ] closes #20353
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -147,7 +147,7 @@ MultiIndex
 I/O
 ^^^
 
--
+- (Only for Windows) Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` (:issue:`20353`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.txt]: I would move the Windows portion as follows:

`Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` on Windows (:issue:`20353`)`

File: pandas/tests/io/formats/test_to_csv.py [hunk 2]
```diff
@@ -285,3 +285,62 @@ def test_to_csv_string_array_utf8(self):
...
+            'str_crlf': ['abc', 'd\r\nef', 'g\r\nh\r\n\r\ni']
+        }
+        df = pd.DataFrame(data)
+
+        with tm.ensure_clean('crlf_test.csv') as path:
+            expected_bin = \
+                b'int,str_crlf\n' \
+                b'1,abc\n' \
+                b'2,"d\r\nef"\n' \
+                b'3,"g\r\nh\r\n\r\ni"\n'
```
Comment (gfyoung) [inline pandas/tests/io/formats/test_to_csv.py]: Use parentheses instead of slashes for multi-line strings.

File: pandas/io/common.py [hunk 3]
```diff
@@ -394,13 +395,13 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
     elif is_path:
         if compat.PY2:
             # Python 2
-            f = open(path_or_buf, mode)
+            f = io.open(path_or_buf, mode, newline='\n')
```
Comment (gfyoung) [inline pandas/io/common.py]: Why do we need to use `io.open` for this one?

Comment (deflatSOCO) [reply pandas/io/common.py]: I referenced [this stackoverflow article](https://stackoverflow.com/questions/34148441/python-2-7-newline-in-function), but it was not appropreate as all CI tests failed in py2 environment.
I will fix it.

File: pandas/io/common.py [hunk 4]
```diff
@@ -394,13 +394,15 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
     elif is_path:
         if compat.PY2:
             # Python 2
+            if mode == 'w':
+                mode = 'wb'
             f = open(path_or_buf, mode)
         elif encoding:
             # Python 3 and encoding
-            f = open(path_or_buf, mode, encoding=encoding)
+            f = open(path_or_buf, mode, encoding=encoding, newline='\n')
```
Comment (bibenb1) [inline pandas/io/common.py]: @deflatSOCO Is there a specific reason for not making newline a function parameter? I would suggest that you make _newline_ a `_get_handle` parameter and default it to *None* then let `open` handle it from there.
Then update `formats.csvs.CSVFormatter.save` to pass _self.line_terminator_ to the function

**Comment by chris-b1** [comment]
Willing to be convinced otherwise but I don't think the default should change to `\n` on Windows.  It's an annoying convention, but convention all the same.

**Comment by gfyoung** [comment]
@chris-b1 : The issue more relates to an inconsistency, as <a href="https://github.com/pandas-dev/pandas/issues/20353#issuecomment-392301505">here</a>.

**Comment by chris-b1** [comment]
Right, and I'm good with fixing the original bug (`\r\n` terminator when `\n` explicitly passed), but this PR would, if I'm understanding correctly, also change the default to `\n` when no line terminator is passed.

**Comment by gfyoung** [comment]
> but this PR would, if I'm understanding correctly, also change the default to \n when no line terminator is passed.

@chris-b1 : Ah, I see what you mean.  That is indeed correct based on the tests.

@deflatSOCO : Could you change it so that the default remains `\r\n` unless otherwise specified?

**Comment by deflatSOCO** [comment]
@chris-b1 @gfyoung  Please let me convince.
[The current document](http://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.to_csv.html) says the default value of line_terminator in `to_csv` is **'\n'** no matter what OS is, so I just followed that in this PR.
I suppose that you mean that should be changed to '\r\n' only in Windows.

If so, the expected output should be like this. Is that right?
* code
```
data = pd.DataFrame({
    "integer":[1,2,3],
    "string_lf":["abc","d\nef","gh\r\ni"]
})
data.to_csv('test.txt', index=False)
print(pd.read_csv('test.csv'))
data.to_csv('test.csv', index=False)
print(pd.read_csv('test.csv'))
with open("test.csv", mode='rb') as f:
    print(f.read())
```
* current output(v0.23)
```
   integer  string_lf
0        1        abc
1        2    d\r\nef
2        3  gh\r\r\ni
b'integer,string_lf\r\n1,abc\r\n2,"d\r\nef"\r\n3,"gh\r\r\ni"\r\n'
```
* current output(this PR)
```
   integer  string_lf
0        1        abc
1        2      d\nef
2        3    gh\r\ni
b'integer,string_lf\n1,abc\n2,"d\nef"\n3,"gh\ni"\n'
```
* expected output
```
   integer  string_lf
0        1        abc
1        2      d\nef
2        3    gh\r\ni
b'integer,string_lf\r\n1,abc\r\n2,"d\nef"\r\n3,"gh\ni"\r\n'
```

**Comment by gfyoung** [comment]
> @chris-b1 @gfyoung Please let me convince.
The current document says the default value of line_terminator in to_csv is '\n' no matter what OS is, so I just followed that in this PR.

@chris-b1 : Thoughts?  I think @deflatSOCO has a point here.

**Comment by gfyoung** [comment]
@jreback @chris-b1 : Could you take a look at this?

**Comment by chris-b1** [comment]
I know the current docs say contrary, but my vote is that we continue defaulting to the platform default line terminator (`\r\n` on Windows) and updating the docs as such.

It preserves backwards compat - so if someone was relying on the `\r` being there, their code won't break, and is consistent with other programs on Windows that save csvs, including Excel, and python's csv writer.

```python
In [1]: import csv

In [3]: with open('tmp.csv', 'wt', newline='') as f:
   ...:     writer = csv.writer(f)
   ...:     writer.writerow(['a', 'b'])
   ...:     writer.writerow([1, 2])
   ...:     writer.writerow([3, 4])

In [4]: open('tmp.csv', 'rb').read()
Out[4]: b'a,b\r\n1,2\r\n3,4\r\n'
```

(again fully in support of merging this with bugfix for case with explicit ending)

**Comment by jreback** [comment]
can u rebase

File: pandas/tests/frame/test_to_csv.py [hunk 5]
```diff
@@ -841,10 +842,10 @@ def test_to_csv_unicodewriter_quoting(self):
...
 
         result = buf.getvalue()
-        expected = ('"A","B"\n'
-                    '1,"foo"\n'
-                    '2,"bar"\n'
-                    '3,"baz"\n')
+        expected = ('"A","B"' + os.linesep +
+                    '1,"foo"' + os.linesep +
+                    '2,"bar"' + os.linesep +
+                    '3,"baz"' + os.linesep)
```
Comment (gfyoung) [inline pandas/tests/frame/test_to_csv.py]: I wonder if we can make a testing utility for this.

File: pandas/tests/frame/test_to_csv.py [hunk 6]
```diff
@@ -865,31 +869,38 @@ def test_to_csv_index_no_leading_comma(self):
...
-        expected = (',A,B\n'
-                    'one,1,4\n'
-                    'two,2,5\n'
-                    'three,3,6\n')
-        assert buf.getvalue() == expected
+            expected = (b',A,B\r\none,1,4\r\ntwo,2,5\r\nthree,3,6\r\n')
+            with open(path, mode='rb') as f:
+                assert f.read() == expected
+
+            df.to_csv(path, line_terminator='\n')
```
Comment (jreback) [inline pandas/tests/frame/test_to_csv.py]: can you do separate ensure_cleans() for each case & put a comment for each case

File: pandas/tests/io/formats/test_to_csv.py [hunk 7]
```diff
@@ -286,21 +342,114 @@ def test_to_csv_string_array_utf8(self):
...
+        os_linesep = os.linesep.encode('utf-8')
+
+        with tm.ensure_clean('lf_test.csv') as path:
+            expected_noarg = (
+                b'int,str_lf' + os_linesep +
+                b'1,abc' + os_linesep +
+                b'2,"d\nef"' + os_linesep +
+                b'3,"g\nh\n\ni"' + os_linesep
+            )
+
```
Comment (jreback) [inline pandas/tests/io/formats/test_to_csv.py]: can you make separate ensure_cleans here

File: pandas/tests/io/formats/test_to_csv.py [hunk 8]
```diff
@@ -286,21 +342,114 @@ def test_to_csv_string_array_utf8(self):
...
+                         '0,foo,bar',
+                         '1,baz,qux']
+        expected_ascii = tm.convert_rows_list_to_csv_str(expected_rows)
         df.to_csv(sys.stdout, encoding='ascii')
         output = sys.stdout.getvalue()
         assert output == expected_ascii
         assert not sys.stdout.closed
 
+    @pytest.mark.xfail(
+        os.name == 'nt',
```
Comment (jreback) [inline pandas/tests/io/formats/test_to_csv.py]: use ``compat.is_windows_platform()``

File: pandas/util/testing.py [hunk 9]
```diff
@@ -2795,3 +2795,10 @@ def skipna_wrapper(x):
             return alternative(nona)
 
     return skipna_wrapper
+
+
+def convert_rows_list_to_csv_str(rows_list):
+    csv_str = ''
```
Comment (jreback) [inline pandas/util/testing.py]: can you add a mini-doc-string

File: pandas/util/testing.py [hunk 10]
```diff
@@ -2795,3 +2795,10 @@ def skipna_wrapper(x):
...
+    for row in rows_list:
+        csv_str += row + os.linesep
+    return csv_str
```
Comment (jreback) [inline pandas/util/testing.py]: can you use 

``return os.linesep.join(rows_list)`` ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/util/testing.py [hunk 11]
```diff
@@ -2801,3 +2801,24 @@ def skipna_wrapper(x):
             return alternative(nona)
 
     return skipna_wrapper
+
+
+def convert_rows_list_to_csv_str(rows_list):
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/util/testing.py]: Feedback was addressed in a715d9c.

**Comment by gfyoung** [comment]
@deflatSOCO : Could you address the merge conflicts here?

@TomAugspurger @jreback : Can you have a look again at this?

**Comment by deflatSOCO** [comment]
@TomAugspurger @jreback : Can you have a look again at this?

File: doc/source/whatsnew/v0.24.0.txt [hunk 12]
```diff
@@ -189,6 +189,7 @@ Other Enhancements
 
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+- :func:`DataFrame.to_csv` now uses :func:`os.linesep` rather than ``'\n'`` for the default line terminator.(:issue:`20353`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you elabortate on what if anything a user needs to do (e.g. when should they care about this)

File: doc/source/whatsnew/v0.24.0.txt [hunk 13]
```diff
@@ -731,6 +732,7 @@ MultiIndex
 I/O
 ^^^
 
+- Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` on Windows (:issue:`20353`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: is this a repeat?

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok to me. i would just request an exapanded whatsnew note. maybe a small sub-section on when a user should care and what they need to do about it. e.g. i write csv's on windows, do I need to do anything?what if I explicity pass ``line_terminator``

File: doc/source/whatsnew/v0.24.0.txt [hunk 14]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _whatsnew_0240.api_breaking.csv_line_terminator:
+
+`os.linesep` is used for ``line_terminator`` of ``DataFrame.to_csv``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`DataFrame.to_csv` now uses :func:`os.linesep` rather than ``'\n'``
+ for the default line terminator(:issue:`20353`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: space after terminator

File: doc/source/whatsnew/v0.24.0.txt [hunk 15]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
...
+- This change only affects when running on Windows, where ``'\r\n'`` was used for line terminator
+even when ``'\n'`` was passed in ``line_terminator``.
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: no need to use the leading '-', just write more in a paragraph style. First sentence is great, not sure I understand the 2nd and 3rd ones.

File: doc/source/whatsnew/v0.24.0.txt [hunk 16]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
...
+  - Strictly speeaking, all ``'\n'``s appear in data and line terminator of CSV were converted into ``'\r\n'``s.
+  - This problem was resolved by passing file object with ``newline='\n'`` option as output, rather than file name.
+
+Previous Behavior on Windows:
+
+.. code-block:: ipython
+
+  In [1]: import pandas as pd
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: you can skip the pd import

File: doc/source/whatsnew/v0.24.0.txt [hunk 17]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
...
+
+  In [4]: print(pd.read_csv("test.csv"))
+      string_with_lf       string_with_crlf
+  0              abc                    abc
+  1          d\r\nef              d\r\r\nef
+  2  g\r\nh\r\n\r\ni  g\r\r\nh\r\r\n\r\r\ni
+
+  In [5]: with open("test.csv", mode='rb') as f:
+     ...:     print(f.read())
+  b'string_with_lf,string_with_crlf\r\nabc,abc\r\n"d\r\nef","d\r\r\nef"\r\n"g\r\nh
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can make this simpler, just show [3] , [4]

File: doc/source/whatsnew/v0.24.0.txt [hunk 18]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
...
+
+- By passing ``line_terminator`` explicitly, line terminator is set to that character.
+- The value of ``line_terminator`` only affects the line terminator of CSV,
+  so it does not change the value inside the data.
+
+.. code-block:: ipython
+
+  In [1]: import pandas as pd
+
+  In [2]: data = pd.DataFrame({
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: same as above

File: doc/source/whatsnew/v0.24.0.txt [hunk 19]
```diff
@@ -196,6 +196,158 @@ Other Enhancements
...
+     ...:     binary_str=f.read()
+     ...: binary_str
+  Out[5]: b'string_with_lf,string_with_crlf\r\nabc,abc\r\n"d\nef","d\r\nef"\r\n"g\
+  nh\n\ni","g\r\nh\r\n\r\ni"\r\n'
+
+  In [6]: with open("test2.csv", mode='w', newline='\n') as f:
+     ...:     data.to_csv(f,index=False,line_terminator='\n')
+
+  In [7]: pd.read_csv("test2.csv")
+  Out[7]:
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: thanks for writing all this, but just show a single, most important change

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 17 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: yes, the point is we want a small example that shows the most common case where a user would want to know what is changing. Too long and you lose the reader, too short and you don't know if the change applies to youl.

File: doc/source/whatsnew/v0.24.0.txt [hunk 20]
```diff
@@ -196,6 +196,94 @@ Other Enhancements
...
+
+Previous Behavior on Windows:
+
+.. code-block:: ipython
+
+In [1]: data = pd.DataFrame({
+   ...: "string_with_lf":["a\nbc"],
+   ...: "string_with_crlf":["a\r\nbc"]
+   ...: })
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you use pep in these examples, e.g. put spaces after the commas

File: doc/source/whatsnew/v0.24.0.txt [hunk 21]
```diff
@@ -196,6 +196,94 @@ Other Enhancements
...
+In [2]: data.to_csv("test.csv",index=False,line_terminator='\n')
+
+In [3]: with open("test.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\r\n"a\r\nbc","a\r\r\nbc"\r\n'
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: I still not clear what these cases are , can you add a coment for example what 2/3 differe from 4/5

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by deflatSOCO** [comment]
@jreback I applied in commit e4badc4, and rebased. Please check them.

File: doc/source/whatsnew/v0.24.0.txt [hunk 22]
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
...
+In [4]: # When passing file OBJECT with newline option to to_csv, line_terminator works.
+   ...: with open("test2.csv", mode='w', newline='\n') as f:
+   ...:     data.to_csv(f, index=False, line_terminator='\n')
+
+In [5]: with open("test2.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\n"a\nbc","a\r\nbc"\n'
+
+
+New Behavior on Windows:
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: @gfyoung can you have a read over this. This should just point out the most common case / change that matters.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt [hunk 23]
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
...
+   ...: })
+
+In [2]: data.to_csv("test.csv", index=False, line_terminator='\n')
+
+In [3]: with open("test.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\n"a\nbc","a\r\nbc"\n'
+
+
+- On windows, the value of ``os.linesep`` is ``'\r\n'``,
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.txt]: windows --> Windows

File: doc/source/whatsnew/v0.24.0.txt [hunk 24]
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
...
+In [5]: with open("test2.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\n"a\nbc","a\r\nbc"\n'
+
+
+New Behavior on Windows:
+
+- By passing ``line_terminator`` explicitly, line terminator is set to that character.
+- The value of ``line_terminator`` only affects the line terminator of CSV,
+  so it does not change the value inside the data.
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.txt]: "it does not change the value inside the data" - what do you mean by that exactly?

File: doc/source/whatsnew/v0.24.0.txt [hunk 25]
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
...
+
+In [2]: data.to_csv("test.csv", index=False)
+
+In [3]: with open("test.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\r\n"a\nbc","a\r\nbc"\r\n'
+
+
+- As default value of ``line_terminator`` changes, just passing file object with ``newline='\n'`` does not set ``'\n'`` to line terminator.
+  Pass ``line_terminator='\n'`` explicitly.
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.txt]: I think we can reword this a little.  Say something like:

~~~
For files objects, specifying `newline` is not sufficient to set the line terminator.  You must pass in the line terminator explicitly, even in this case.
~~~

File: doc/source/whatsnew/v0.24.0.txt (hunk 22 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.24.0.txt]: Overall, this looks pretty good!  Left a few comments below.

File: doc/source/whatsnew/v0.24.0.txt (hunk 24 above)
Comment (deflatSOCO) [reply doc/source/whatsnew/v0.24.0.txt]: Suppose that `data` is like this (which is used in following codes)

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1|a\nbc|a\r\nbc|

As described in [2]&[3] of "Previous Behavior", if there are some `\n`s INSIDE the data, `to_csv()` changed the value and outputted CSV like this.

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1| a **\r** \nbc | a **\r** \r\nbc |

By this PR, it is fixed to output the following CSV.

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1| a\nbc | a\r\nbc |

Comment (gfyoung) [reply doc/source/whatsnew/v0.24.0.txt]: Gotcha.  Could you mention that in [2] and [3]?  What you're describing isn't explicitly mentioned.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@gfyoung you ok with this, can you rebase this? or @deflatSOCO

**Comment by gfyoung** [comment]
@jreback : Can do, given that no reply how has come since the latest comment in > 1 week.

**Comment by gfyoung** [comment]
Alas, there were too many merge conflicts with `master` that I had to nuke the old commits and replace with it a squashed version.

**Comment by gfyoung** [comment]
@jreback : Addressed all comments, rebased, and all is green.  PTAL.

**Comment by jreback** [comment]
thanks @deflatSOCO and @gfyoung

**Pull request merged**